nomos-mix/core/src/message_blend/temporal.rs

```diff
@@ -1,8 +1,101 @@
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{Future, Stream};
+use rand::Rng;
 use serde::{Deserialize, Serialize};
+use tokio::time;
+
+/// [`TemporalProcessor`] delays messages randomly to hide timing correlation
+/// between incoming and outgoing messages from a node.
+///
+/// See the [`Stream`] implementation below for more details on how it works.
+pub struct TemporalProcessor<M> {
+    settings: TemporalProcessorSettings,
+    // All scheduled messages
+    queue: VecDeque<M>,
+    /// Interval in seconds for running the lottery to release a message
+    lottery_interval: time::Interval,
+    /// To wait a few seconds after running the lottery before releasing the message.
+    /// The lottery returns how long to wait before releasing the message.
+    release_timer: Option<Pin<Box<time::Sleep>>>,
+}
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct TemporalProcessorSettings {
     pub max_delay_seconds: u64,
 }
 
-// TODO: Implement TemporalProcessor
+impl<M> TemporalProcessor<M> {
+    #[allow(dead_code)] // TODO: Remove this macro once this module is integrated
+    pub fn new(settings: TemporalProcessorSettings) -> Self {
+        let lottery_interval = Self::lottery_interval(settings.max_delay_seconds);
+        Self {
+            settings,
+            queue: VecDeque::new(),
+            lottery_interval,
+            release_timer: None,
+        }
+    }
+
+    /// Create [`time::Interval`] for running the lottery to release a message.
+    fn lottery_interval(max_delay_seconds: u64) -> time::Interval {
```

Comment:
If they are seconds I think it would be better tu have the type as Duration (?)

Reply:
It was because I wanted to explicitly have it as a "second", which doesn't have a fractional point, because this parameter is highly related to the `max_emission_frequency: f64`, which means the max num of messages per "second".